### PR TITLE
Improve handling of Ubuntu series and OS release combinations

### PR DESCRIPTION
**Pull request by agileshaw**
- Get current minimum series of the cloud in the analysis part
- set current_cloud_os_release and current_cloud_series right away in the analysis part so that errors can be caught early
- Add LTS_TO_OS_RELEASE, which maps a LTS series to a range of supporting OS releases
- Add function in plan stage to determine the upgrade route based on series/release combination, including limiting the support range of COU to focal ussuri - yoga
- Add message about which release we're upgrading from and to
- Fix release codename typo: kinect -> kinetic

fixes: #99 
fixes: #129

### CODE REVIEW

File: tests/unit/steps/test_steps_plan.py [hunk 1]
```diff
@@ -162,14 +162,74 @@ async def test_generate_plan(apps, model):
...
-async def test_generate_plan_raise_NoTargetError(mocker):
-    exp_error_msg = "Cannot find target to upgrade."
-    analysis_result = MagicMock(spec=Analysis)
-    analysis_result.current_cloud_os_release.next_release = None
-    # not possible to determine target
+@pytest.mark.parametrize(
+    "current_os_release, current_series, next_release",
+    [
+        (OpenStackRelease("victoria"), "focal", "wallaby"),
+        (OpenStackRelease("yoga"), "jammy", "zed"),
```
Comment (gabrielcocenza) [inline tests/unit/steps/test_steps_plan.py:169]: This test should fail IMO. If a user manually upgrade the series to jammy and run COU I would fell more comfortable not letting go ahead and upgrade to zed.

**Review by gabrielcocenza — CHANGES_REQUESTED** [review]

File: cou/steps/analyze.py [hunk 2]
```diff
@@ -43,6 +43,11 @@ class Analysis:
     apps_control_plane: list[OpenStackApplication]
     apps_data_plane: list[OpenStackApplication]
 
+    def __post_init__(self) -> None:
+        """Initialize the Analysis dataclass."""
+        self.current_cloud_os_release = self._get_minimum_cloud_os_release()
+        self.current_cloud_series = self._get_minimum_cloud_series()
```
Comment (rgildein) [inline cou/steps/analyze.py:49]: Can we define these in dataclass itself and set value only here? e.g.
 ```python
>>> @dataclass
... class A:
...  a: str = field(init=False)
...  def __post_init__(self) -> None:
...   self.a = "test"
... 
>>> obj = A()
>>> obj.a
'test'
```

File: cou/steps/plan.py [hunk 3]
```diff
@@ -117,3 +119,55 @@ async def create_upgrade_group(
...
+    :return: The target OS release to upgrade the cloud to.
+    :rtype: str
+    """
+    if not current_os_release:
+        raise NoTargetError(
+            "Cannot determine the current OS release in the cloud. "
+            "Is this a valid OpenStack cloud?"
+        )
+
+    supporting_lts_series = ",".join(list(LTS_TO_OS_RELEASE.keys()))
```
Comment (rgildein) [inline cou/steps/plan.py:144]: not-blocking: 
```suggestion
    supporting_lts_series = ",".join(LTS_TO_OS_RELEASE)
```

File: cou/steps/plan.py [hunk 4]
```diff
@@ -117,3 +119,55 @@ async def create_upgrade_group(
...
+
+    try:
+        # Check if the release is the "last" supported by the series
+        if str(current_os_release) == LTS_TO_OS_RELEASE[current_series][-1]:
+            print(
+                f"No upgrade available. '{current_os_release}' (the minimum OS release found "
+                f"in the cloud) is the highest release '{current_series}' (the minimum Ubuntu "
+                "series in the cloud) supports."
+            )
+            sys.exit(0)
```
Comment (rgildein) [inline cou/steps/plan.py:159]: To me this is not an good idea hiding sys exit in code. I think we should raise and exception here. Exception should take that string as message and in CLI we should catch it and print the message.

File: tests/unit/steps/test_steps_analyze.py [hunk 5]
```diff
@@ -137,6 +139,30 @@ async def test_analysis_detect_current_cloud_os_release_same_release(apps, model
     assert result.current_cloud_os_release == "ussuri"
 
 
+@pytest.mark.asyncio
+async def test_analysis_detect_current_cloud_series_same_series(apps, model):
+    result = analyze.Analysis(
+        model=model,
+        apps_control_plane=[apps["rmq_ussuri"], apps["keystone_wallaby"], apps["cinder_ussuri"]],
+        apps_data_plane=[],
```
Comment (rgildein) [inline tests/unit/steps/test_steps_analyze.py:147]: I think that we should used mocked apps here, this is not related with app class at all and if that failed all tests like this will failed and that bad practice.

**Review by rgildein — CHANGES_REQUESTED** [review]

File: cou/steps/plan.py [hunk 6]
```diff
@@ -117,3 +119,55 @@ async def create_upgrade_group(
...
+            f"The supporting series are: {supporting_lts_series}"
+        )
+
+    try:
+        # Check if the release is the "last" supported by the series
+        if str(current_os_release) == LTS_TO_OS_RELEASE[current_series][-1]:
+            print(
+                f"No upgrade available. '{current_os_release}' (the minimum OS release found "
+                f"in the cloud) is the highest release '{current_series}' (the minimum Ubuntu "
+                "series in the cloud) supports."
```
[the remark this answers is not in the document]
Comment (agileshaw) [reply cou/steps/plan.py]: I re-worded this message. Please let me know if it is clearer

File: cou/steps/plan.py [hunk 7]
```diff
@@ -117,3 +119,55 @@ async def create_upgrade_group(
...
+            )
+            sys.exit(0)
+
+        # get the next release as the target from the current cloud os release
+        target = current_os_release.next_release
+        if not target:
```
[the remark this answers is not in the document]
Comment (agileshaw) [reply cou/steps/plan.py:174]: The purpose of this PR was to just have a smart way of determine whether the next move should be series upgrade or release upgrade, so limiting the support range to ussuri<->yoga was not in scope. But I think we could expand the scope of this PR and include it here. Updated in the new patch

File: cou/steps/plan.py [hunk 8]
```diff
@@ -117,3 +119,55 @@ async def create_upgrade_group(
...
+            raise NoTargetError(
+                "Cannot find target to upgrade. Current minimum OS release is "
+                f"'{str(current_os_release)}'. Current Ubuntu series is '{current_series}'."
```
[the remark this answers is not in the document]
Comment (agileshaw) [reply cou/steps/plan.py:177]: I'm also letting the user know the starting point should be larger than ussuri

File: tests/unit/steps/test_steps_plan.py (hunk 1 above)
Comment (agileshaw) [reply tests/unit/steps/test_steps_plan.py]: As we are including the range limit, I revised the tests

File: cou/steps/analyze.py (hunk 2 above)
Comment (agileshaw) [reply cou/steps/analyze.py:51]: Good suggestion. Thanks :+1:

File: cou/steps/plan.py (hunk 4 above)
Comment (agileshaw) [reply cou/steps/plan.py]: I was actually torn between this approach and what you are suggesting here. After re-consideration, yes, I think putting all `sys.exit` in `cli.py` is better.

File: tests/unit/steps/test_steps_analyze.py (hunk 5 above)
Comment (agileshaw) [reply tests/unit/steps/test_steps_analyze.py:147]: We could but I don't think it is necessary because this is defined in `conftest.py`, which is meant to be shared across multiple test modules.

File: cou/steps/plan.py [hunk 9]
```diff
@@ -119,3 +128,66 @@ async def create_upgrade_group(
...
+        if current_os_release < OpenStackRelease("ussuri") or OpenStackRelease(
+            target
+        ) > OpenStackRelease("yoga"):
+            raise ReleaseOutOfRange(
+                "At the moment COU only supports upgrade in the range of `ussuri` and `yoga` "
+                "(inclusive) for `focal` series. Current minimum OS release is "
+                f"'{str(current_os_release)}'. Current Ubuntu series is '{current_series}'."
+            )
+        return target
+    except KeyError:  # raise exception if the current series is not in LTS_TO_OS_RELEASE map
```
Comment (gabrielcocenza) [inline cou/steps/plan.py:189]: IMO this try and except `KeyError` is catching too much code. Is  it possible to isolate where you are really checking that?

**Comment by agileshaw** [comment]
Create issue #144 to remind us to expand supporting range when COU can handle more releases/series in the future

**Review by rgildein — APPROVED** [review]
LGTM

**Review by gabrielcocenza — APPROVED** [review]
